rs/registry/canister/canister/canister.rs

```diff
@@ -185,60 +185,63 @@ ic_nervous_system_common_build_metadata::define_get_build_metadata_candid_method
 
 #[export_name = "canister_query get_changes_since"]
 fn get_changes_since() {
-    // Only self-authenticating and anonymous principals are allowed to call us.
-    const ALLOWED_CALLER_CLASSES: [Result<PrincipalIdClass, String>; 2] = [
-        Ok(PrincipalIdClass::SelfAuthenticating),
-        Ok(PrincipalIdClass::Anonymous),
-    ];
-    let class = caller().class();
-    if !ALLOWED_CALLER_CLASSES.contains(&class) {
-        let response = RegistryGetChangesSinceResponse {
-            error: Some(RegistryError {
-                code: Code::Authorization as i32,
-                reason: format!(
+    fn main() -> Result<RegistryGetChangesSinceResponse, (Code, String)> {
+        // Authorize: Only self-authenticating and anonymous principals are allowed to call us.
+        const ALLOWED_CALLER_CLASSES: [Result<PrincipalIdClass, String>; 2] = [
+            Ok(PrincipalIdClass::SelfAuthenticating),
+            Ok(PrincipalIdClass::Anonymous),
+        ];
+        let class = caller().class();
+        if !ALLOWED_CALLER_CLASSES.contains(&class) {
+            return Err((
+                Code::Authorization,
+                format!(
                     "Caller must be self-authenticating, or anonymous (but was {:?}).",
                     class,
                 ),
-                key: vec![],
-            }),
-            ..Default::default()
-        };
+            ));
+        }
 
-        let response: Vec<u8> =
-            serialize_get_changes_since_response(response).expect("Error serializing response");
+        // Parse request.
+        let request = deserialize_get_changes_since_request(arg_data())
+            .map_err(|err| (Code::MalformedMessage, err.to_string()))?;
+        let version = request;
 
-        reply(&response);
-        return;
-    }
+        // All requirements met. Proceed with "real work".
 
-    let response_pb = match deserialize_get_changes_since_request(arg_data()) {
-        Ok(version) => {
-            let registry = registry();
+        let registry = registry();
 
-            let max_versions = registry
-                .count_fitting_deltas(version, MAX_REGISTRY_DELTAS_SIZE)
-                .min(MAX_VERSIONS_PER_QUERY);
+        let max_versions = registry
+            .count_fitting_deltas(version, MAX_REGISTRY_DELTAS_SIZE)
+            .min(MAX_VERSIONS_PER_QUERY);
+
+        Ok(RegistryGetChangesSinceResponse {
```

Comment:
NCR: Minor thing, but I might have kept a symmetry around the Err/Ok not being the response object itself, and instead create the response object below?  

Random thoughts:
This pattern sort of reminds me of normal web applications with an API layer, where the API layer translates the request into internal operations (ideally in a simple way) and then translates that back into a response.  It's largely what we've been working on doing with our API decoupling.  

It makes me wonder a bit if maybe some of what's happening in this function ought to live somewhere else anyway.